polygeist/test/polygeist-opt/bareptrlowering.mlir

```diff
@@ -0,0 +1,553 @@
+
+// -----
+
+// CHECK-LABEL:   llvm.func @ptr_args_and_ret(!llvm.ptr<i64>, !llvm.ptr<i64>) -> !llvm.ptr<i64>
+
+func.func private @ptr_args_and_ret(%arg0: memref<1xi64>, %arg1: memref<?xi64>) -> memref<?x4x4xi64>
+
+// -----
+
+// CHECK-LABEL:   llvm.func @ptr_args_and_ret_with_attrs(!llvm.ptr<i64> {llvm.byval = !llvm.ptr<i64>}, !llvm.ptr<i64> {llvm.byval = !llvm.ptr<i64>}) -> !llvm.ptr<i64>
+
+func.func private @ptr_args_and_ret_with_attrs(%arg0: memref<1xi64> {llvm.byval = memref<1xi64>},
+                                               %arg1: memref<?xi64> {llvm.byval = memref<?xi64>}) -> memref<?x4x4xi64>
+
+// -----
+
+gpu.module @kernels {
+
+// CHECK-LABEL:   llvm.func @kernel(
+// CHECK-SAME:                      %[[VAL_0:.*]]: !llvm.ptr<i64> {llvm.byval = !llvm.ptr<i64>},
+// CHECK-SAME:                      %[[VAL_1:.*]]: !llvm.ptr<i64> {llvm.byval = !llvm.ptr<i64>}) attributes {gpu.kernel, workgroup_attributions = 0 : i64} {
+// CHECK-NEXT:      llvm.return
+// CHECK-NEXT:         }
+
+  gpu.func @kernel(%arg0: memref<1xi64> {llvm.byval = memref<1xi64>},
+                   %arg1: memref<?xi64> {llvm.byval = memref<?xi64>}) kernel {
+    gpu.return
+  }
+}
+
+// -----
+
+// CHECK-LABEL:   llvm.mlir.global external @global() {addr_space = 0 : i32} : !llvm.array<3 x i64>
+
+memref.global @global : memref<3xi64>
+
+// CHECK-LABEL:   llvm.func @get_global() -> !llvm.ptr<i64>
+// CHECK-NEXT:      %[[VAL_0:.*]] = llvm.mlir.addressof @global : !llvm.ptr<array<3 x i64>>
+// CHECK-NEXT:      %[[VAL_1:.*]] = llvm.getelementptr %[[VAL_0]][0, 0] : (!llvm.ptr<array<3 x i64>>) -> !llvm.ptr<i64>
+// CHECK-NEXT:      llvm.return %[[VAL_1]] : !llvm.ptr<i64>
+// CHECK-NEXT:    }
+
+func.func private @get_global() -> memref<3xi64> {
+  %0 = memref.get_global @global : memref<3xi64>
+  return %0 : memref<3xi64>
+}
+
+// -----
+
+// CHECK-LABEL:   llvm.mlir.global external @global_addrspace() {addr_space = 4 : i32} : !llvm.array<3 x i64>
+
+memref.global @global_addrspace : memref<3xi64, 4>
+
+// CHECK-LABEL:   llvm.func @get_global_addrspace() -> !llvm.ptr<i64, 4>
+// CHECK-NEXT:      %[[VAL_0:.*]] = llvm.mlir.addressof @global_addrspace : !llvm.ptr<array<3 x i64>, 4>
+// CHECK-NEXT:      %[[VAL_1:.*]] = llvm.getelementptr %[[VAL_0]][0, 0] : (!llvm.ptr<array<3 x i64>, 4>) -> !llvm.ptr<i64, 4>
+// CHECK-NEXT:      llvm.return %[[VAL_1]] : !llvm.ptr<i64, 4>
+// CHECK-NEXT:    }
+
+func.func private @get_global_addrspace() -> memref<3xi64, 4> {
+  %0 = memref.get_global @global_addrspace : memref<3xi64, 4>
+  return %0 : memref<3xi64, 4>
+}
+
+// -----
+
+memref.global "private" constant @shape : memref<2xi64> = dense<[2, 2]>
+
+// CHECK-LABEL:   llvm.func @reshape(
+// CHECK-SAME:                       %[[VAL_0:.*]]: !llvm.ptr<i32>) -> !llvm.ptr<i32>
+// CHECK:           %[[VAL_2:.*]] = llvm.getelementptr %{{.*}}[0, 0] : (!llvm.ptr<array<2 x i64>>) -> !llvm.ptr<i64>
+// CHECK-NEXT:      llvm.return %[[VAL_0]] : !llvm.ptr<i32>
```

Comment:
`reshape` simply changes the shape of a memref. In our new lowering, we do not care about shapes (nor layouts), as we impose restrictions (identity layout and static dimensions in the range `[1, n)`) so that memory accesses can be performed with no additional information (just the pointer and the indices on each dimension). As the `reshape` operation does not alter the underlying pointer and we ensure the restrictions are still held on output, we can just make `reshape` the identity operation in our lowering.

Reply:
Yes I actually saw the lowering code for the reshape operation. I guess the getelementptr operation will be cleaned up by `opt` given it is without uses.